Added reference to RFC2818 along with guidance to add subjectAltNames…

JeffH-AWS opened the pull request:
Fixes #730 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

JeffH-AWS commented on _security-plugin/configuration/generate-certificates.md:
```diff
@@ -88,7 +88,7 @@ Just like the root certificate, use the `-days` option to specify an expiration
 Follow the steps in [Generate an admin certificate](#generate-an-admin-certificate) with new file names to generate a new certificate for each node and as many client certificates as you need. For example, you might generate one client certificate for OpenSearch Dashboards and another for a Python client. Each certificate should use its own private key.
 
-If you generate node certificates and have `plugins.security.ssl.transport.enforce_hostname_verification` set to `true` (default), be sure to specify a common name (CN) for the certificate that matches the hostname of the intended node. If you want to use the same node certificate on all nodes (not recommended), set hostname verification to `false`. For more information, see [Configure TLS certif ...
+If you generate node certificates and have `plugins.security.ssl.transport.enforce_hostname_verification` set to `true` (default), be sure to specify a common name (CN) for the certificate that matches the hostname of the intended node. Additionally, we recommend specifying a Subject Alternative Name (SAN) to ensure compliance with [RFC 2818 (HTTP Over TLS)](https://datatracker.ietf.org/doc/html/ ...
```
I'll clean this part up.

setiah commented on _security-plugin/configuration/generate-certificates.md:
```diff
@@ -88,7 +88,7 @@ Just like the root certificate, use the `-days` option to specify an expiration
 Follow the steps in [Generate an admin certificate](#generate-an-admin-certificate) with new file names to generate a new certificate for each node and as many client certificates as you need. For example, you might generate one client certificate for OpenSearch Dashboards and another for a Python client. Each certificate should use its own private key.
 
-If you generate node certificates and have `plugins.security.ssl.transport.enforce_hostname_verification` set to `true` (default), be sure to specify a common name (CN) for the certificate that matches the hostname of the intended node. If you want to use the same node certificate on all nodes (not recommended), set hostname verification to `false`. For more information, see [Configure TLS certif ...
+If you generate node certificates and have `plugins.security.ssl.transport.enforce_hostname_verification` set to `true` (default), be sure to specify a common name (CN) for the certificate that matches the hostname of the intended node. Additionally, we recommend specifying a Subject Alternative Name (SAN) to ensure compliance with [RFC 2818 (HTTP Over TLS)](https://datatracker.ietf.org/doc/html/ ...
```
this looks good. Here's additional context from the RFC

"If a subjectAltName extension of type dNSName is present, that MUST
   be used as the identity. Otherwise, the (most specific) Common Name
   field in the Subject field of the certificate MUST be used. Although
   the use of the Common Name is existing practice, it is deprecated and
   Certification Authorities are encouraged to use the dNSName instead."

setiah reviewed (commented):
Highly recommend someone from security dev team testing this before adding this to documentation, to make sure we don't break the setup experience.

setiah commented on _security-plugin/configuration/generate-certificates.md:
```diff
@@ -99,26 +99,26 @@ If you already know the certificate details and don't want to specify them inter
 openssl genrsa -out admin-key-temp.pem 2048
 openssl pkcs8 -inform PEM -outform PEM -in admin-key-temp.pem -topk8 -nocrypt -v1 PBE-SHA1-3DES -out admin-key.pem
-openssl req -new -key admin-key.pem -subj "/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=ADMIN" -out admin.csr
+openssl req -new -key admin-key.pem -subj "/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=ADMIN" -addext "subjectAltName = DNS:admin.dns.a-record" -out admin.csr
```
Not sure what's the right expectation when the CN is not a dns name. Should we make this a DNS name instead? @mpfz0r thoughts?

mpfz0r replied:
For consistency, I'd use the same DNS name for the CN as for the subjectAltName. 
In practice, I think the CN will mostly be ignored if a subjectAltName is present.

Thank you for fixing this :pray:

JeffH-AWS replied:
Thanks for the guidance @mpfz0r!  I will update accordingly.

On a side note, while I was testing this process today in my dev environment I was running into an error that `-addext "subjectAltName = DNS:admin.dns.a-record"` was not a valid option.  Any opposition to recommending that the user simply add the SAN info in an extension and then pass it to the `openssl` command with `-extfile`?  So the command to generate the signed cert would look something like:

`openssl x509 -req -in node.csr -CA ca.pem -CAkey ca.key -out node.crt -days 365 -sha256 -extfile node.ext`

I use my own root cert to sign my dev certs so I know the syntax is a little different than what we would use in the docs.

mpfz0r replied:
@JeffH-AWS yeah that's quite a mess. see https://security.stackexchange.com/questions/74345/provide-subjectaltname-to-openssl-directly-on-the-command-line 
for all the gory details.

There's multiple solutions, but I think using `-extfile` is the easiest to understand.
Maybe add a line that generates it first?
e.g. `echo 'subjectAltName=DNS:admin.dns.a-record' > node.ext`

JeffH-AWS replied:
Perfect!  I think that sounds good.  I was also considering that the docs advise use of a self-signed cert, although we could very easily add a couple steps for creating a root private key and cert, use the self-created root key/cert to sign all the requests for the cluster, and then add trust for the root cert to all the nodes and any clients that interact with the cluster.

JeffH-AWS replied:
That may be a bit bigger of a lift which I can tackle in another issue.  For now I went ahead and applied the `-extfile` method for simplicity.

cliu123 commented on _security-plugin/configuration/generate-certificates.md:
```diff
@@ -99,26 +99,26 @@ If you already know the certificate details and don't want to specify them inter
 openssl genrsa -out node1-key-temp.pem 2048
 openssl pkcs8 -inform PEM -outform PEM -in node1-key-temp.pem -topk8 -nocrypt -v1 PBE-SHA1-3DES -out node1-key.pem
-openssl req -new -key node1-key.pem -subj "/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=node1.example.com" -out node1.csr
+openssl req -new -key node1-key.pem -subj "/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=node1.example.com" -addext "subjectAltName = DNS:node1.dns.a-record" -out node1.csr
```
Agree! Using the DNS name same as CN for `subjectAltName` will be more accurate.

JeffH-AWS replied:
Thank you!  The change has been made and applied consistently to the doc.

JeffH-AWS commented:
Should be good.  Running the sample script is successful:


Running the sample script locally to test for syntax errors...
```
$ ./sample_script.sh 
Generating RSA private key, 2048 bit long modulus
....+++
.................+++
e is 65537 (0x10001)
Generating RSA private key, 2048 bit long modulus
..................................................................+++
...+++
e is 65537 (0x10001)
Signature ok
subject=/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=admin.dns.a-record
Getting CA Private Key
Generating RSA private key, 2048 bit long modulus
........................................................+++
...............+++
e is 65537 (0x10001)
Signature ok
subject=/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=node1.dns.a-record
Getting CA Private Key
Generating RSA private key, 2048 bit long modulus
.................................................+++
...........................+++
e is 65537 (0x10001)
Signature ok
subject=/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=node2.dns.a-record
Getting CA Private Key
Generating RSA private key, 2048 bit long modulus
.............................+++
.+++
e is 65537 (0x10001)
Signature ok
subject=/C=CA/ST=ONTARIO/L=TORONTO/O=ORG/OU=UNIT/CN=client.dns.a-record
Getting CA Private Key
```

The expected files are created and preserved after the sample script completes
```
$ ll
-rw-r--r--  1 user  group  1704 Jul 11 14:49 admin-key.pem
-rw-r--r--  1 user  group  1265 Jul 11 14:49 admin.pem
-rw-r--r--  1 user  group  1704 Jul 11 14:49 client-key.pem
-rw-r--r--  1 user  group  1269 Jul 11 14:49 client.pem
-rw-r--r--  1 user  group  1704 Jul 11 14:49 node1-key.pem
-rw-r--r--  1 user  group  1265 Jul 11 14:49 node1.pem
-rw-r--r--  1 user  group  1704 Jul 11 14:49 node2-key.pem
-rw-r--r--  1 user  group  1265 Jul 11 14:49 node2.pem
-rw-r--r--  1 user  group  1675 Jul 11 14:49 root-ca-key.pem
-rw-r--r--  1 user  group  1208 Jul 11 14:49 root-ca.pem
-rw-r--r--  1 user  group    17 Jul 11 14:49 root-ca.srl
-rwxr-xr-x  1 user  group  2389 Jul 11 14:49 sample_script.sh
```

Spot checking for accuracy...
```
$ openssl x509 -in node2.pem -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 9476560323131516286 (0x83838167a8d0a57e)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: C=CA, ST=ONTARIO, L=TORONTO, O=ORG, OU=UNIT, CN=root.dns.a-record
        Validity
            Not Before: Jul 11 21:49:43 2022 GMT
            Not After : Jul 10 21:49:43 2024 GMT
        Subject: C=CA, ST=ONTARIO, L=TORONTO, O=ORG, OU=UNIT, CN=node2.dns.a-record
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:ce:d4:9d:0f:71:3f:fa:cf:24:0f:43:44:cb:80:
                    34:a7:31:e7:c0:33:39:d5:f6:4c:38:67:a3:c8:9b:
                    0c:ca:e2:94:4c:6e:99:10:1e:6a:58:77:78:68:5e:
                    ed:e0:91:49:f5:d9:53:8e:3c:1b:3c:cb:53:c3:07:
                    8a:c7:ed:46:27:5d:a4:66:1b:d2:d5:35:2f:54:1e:
                    83:60:10:ae:e1:6c:48:32:c7:4d:0d:fc:07:fd:55:
                    41:00:dc:92:9c:bf:a1:2a:94:35:15:9f:2f:a5:6d:
                    f1:51:85:8c:0a:dd:59:02:85:70:47:a2:ad:10:4e:
                    b3:11:e5:ab:15:c0:11:0e:4b:8e:ce:70:8c:94:b9:
                    34:6a:49:3f:86:d5:68:38:f6:18:1e:e4:35:00:03:
                    4b:4b:69:30:01:5f:35:14:3b:a2:cd:1a:9c:7b:76:
                    91:96:55:5a:2f:41:5d:e1:1d:f8:e6:7c:9e:dd:38:
                    fe:b2:9a:ae:06:aa:5f:f9:46:ae:3f:52:ab:d1:7b:
                    cf:ce:96:f1:f7:f6:cd:76:c5:2e:fa:58:ba:4e:79:
                    b2:f2:30:a0:bc:72:46:5b:d1:ac:bb:12:df:4c:50:
                    81:f1:b9:01:fe:f3:5a:84:d8:62:8f:a1:f8:b2:13:
                    db:8d:56:f3:8a:d7:59:32:ff:fb:a6:01:9a:d6:ec:
                    a2:43
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Subject Alternative Name: 
                DNS:node2.dns.a-record
    Signature Algorithm: sha256WithRSAEncryption
         66:93:5c:17:32:55:aa:6a:da:a6:4e:68:dc:7e:cc:d1:8f:c4:
         9b:14:96:1e:64:8d:e5:36:d2:45:19:b8:c0:f4:bc:3f:db:bf:
         60:12:87:61:de:99:28:03:6e:d1:77:7e:f3:fa:1b:98:61:6d:
         0f:38:63:d7:9d:2a:e7:4b:2f:fc:5f:03:bc:0c:d9:cb:79:f7:
         62:a9:04:2f:a4:ad:51:78:87:cb:90:ad:00:1b:17:a6:a2:dd:
         c0:45:60:52:24:63:88:13:9b:10:78:78:cb:0c:5c:0b:dc:c5:
         65:46:81:8b:4a:f1:c4:91:25:8c:ea:f4:20:10:15:11:1d:f0:
         7a:a3:68:cd:95:29:0c:f0:2e:19:41:e2:96:12:98:05:96:83:
         1c:8e:4f:65:03:a4:69:0a:e4:26:4f:74:d9:d6:fc:f5:a6:fe:
         71:14:69:74:03:97:f2:69:f5:92:60:cf:b5:86:5b:e6:66:7f:
         98:83:5c:d4:a0:f1:bb:07:d4:a4:22:43:aa:b5:40:7e:b0:62:
         b6:7a:23:0e:da:96:32:2a:25:5e:db:ce:71:23:eb:25:ae:89:
         b2:29:1e:30:41:7b:d5:e0:8b:0c:e8:af:9e:d6:9a:43:27:22:
         00:2c:fc:00:18:a3:f5:ff:25:dc:c9:17:01:1e:9e:3d:4d:51:
         4c:91:60:28
```

JeffH-AWS commented:
Ok I did some major clean up.  All CN and SAN examples use the format `example-hostname.dns.a-record`.  I also modified the sample script to include the SAN extension files and remove them upon completion, along with some commentary about the importance of matching CN with SAN.  I do believe this is ready for a final tech review when someone has a chance.  Please let me know if you have any further recommended changes and thank you all so much for being so responsive!!

setiah commented on _security-plugin/configuration/generate-certificates.md:
```diff
@@ -75,20 +75,31 @@ openssl req -new -key admin-key.pem -out admin.csr
-Finally, generate the certificate itself:
+If you generate TLS certificates and have enabled hostname verification by setting `plugins.security.ssl.transport.enforce_hostname_verification` to `true` (default), be sure to specify a common name (CN) for each certificate signing request (CSR) that matches the corresponding DNS A record of the intended node.
+
+If you want to use the same node certificate on all nodes (not recommended), set hostname verification to `false`. For more information, see [Configure TLS certificates]({{site.url}}{{site.baseurl}}/security-plugin/configuration/tls#advanced-hostname-verification-and-dns-lookup).
```
Not sure about recommending hostname verification to false. @cliu123 care to comment if this looks okay from security perspective?

JeffH-AWS replied:
Yeah I am curious about this as well.  It was already existing in the docs.  The setting is described at https://opensearch.org/docs/latest/security-plugin/configuration/tls#advanced-hostname-verification-and-dns-lookup

Perhaps as an alternative I could reword line 80 to read something along the lines of:

```
To learn more about settings related to TLS certificates in OpenSearch, see [Configure TLS certificates]({{site.url}}{{site.baseurl}}/security-plugin/configuration/tls#advanced-hostname-verification-and-dns-lookup).
```

cliu123 replied:
> Not sure about recommending hostname verification to false. @cliu123 care to comment if this looks okay from security perspective?

It's fine to recommend hostname verification to false. Actually this is the only choice in this case, otherwise, verification would fail and error would be thrown. This is not a recommanded use case, which has already been noted as `(not recommended)`. So the descritpion makes sense to me.

setiah reviewed (approved):
one minor comment for security team. Looks good otherwise.

kolchfa-aws reviewed (approved):
LGTM

JeffH-AWS commented:
I applied a few additional changes.

- Removed mention of SAN extension from the admin cert since that's completely extraneous and makes no practical sense.
- Expanded description of host/client search. Rather than directing the user to follow admin cert steps I expanded them since the host/client certs require the additional step of defining a SAN extension
- Cleaned up some phrasing
- Adjusted the sample script to reflect the other changes

kolchfa-aws commented on _security-plugin/configuration/generate-certificates.md:
````diff
@@ -86,9 +90,40 @@ Just like the root certificate, use the `-days` option to specify an expiration
+echo 'subjectAltName=DNS:node1.dns.a-record' > node1.ext
+```
+
+Now generate the certificate:
````
I would use "then" instead of "now"
